<a href="https://colab.research.google.com/github/chrisdionisius/MachineLearning2022/blob/main/Jobsheet_12_Tunning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Jobsheet 12 - Tunning
- NIM   : 2241727008
- Nama  : Dionisius Damarta Yapenrui
- Kelas : TI-4J

## Tugas

1. Menggunakan data (WBC) dan alogirtma yang sama (DT) pada jobsheet js11_02, silahkan implemnetasikan ulang teknik *exhausive search* dengan menggunakan ***random search***. Scikit-learn sudah menyediakan fungsi ini.

2. Menggunakan data yang sama, lakukan tunning dengan menggunakan algoritma SVM. Hyperparamter yang akan ditunning adalah,
    - kernel (linier, polinomial, rbf)
    - degree (akan berdampak jika polinomial)
    - koefisien C (regularization)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# baca data
data = pd.read_csv('/content/drive/MyDrive/dataset/wbc.csv')
data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [4]:
# slice kolom terakhir --> Unnamed column
data = data.iloc[:,:-1]

data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [5]:
X = data.iloc[:, 2:]
y = data['diagnosis']

X.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [6]:
from sklearn.model_selection import RepeatedKFold

# inisiasi repated k-fold
cv = RepeatedKFold(n_splits=4, n_repeats=3, random_state=42)

In [7]:
# from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble  import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform, truncnorm, randint

# inisiasi model
dt = RandomForestClassifier()

# Definisikan hyperparameter yang akan digunakan
params = {
    # randomly sample numbers from 4 to 204 estimators
    'n_estimators': randint(4,200),
    # normally distributed max_features, with mean .25 stddev 0.1, bounded between 0 and 1
    'max_features': truncnorm(a=0, b=1, loc=0.25, scale=0.1),
    # uniform distribution from 0.01 to 0.2 (0.01 + 0.199)
    'min_samples_split': uniform(0.01, 0.199)
}

# inisiasi randomized search berdasarkan nilai repeated k-fold dan hyperparameter
rsc = RandomizedSearchCV(dt,params, cv=cv, n_iter=100, random_state=1)

# train the random search meta-estimator to find the best model out of 100 candidates
model = rsc.fit(X, y)

# Evaluasi dengan score
score = model.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {rsc.best_params_}')

Hasil evaluasi: 0.9947275922671354
Konfigurasi hyperparameter: {'max_features': 0.25411986663039127, 'min_samples_split': 0.013158723180134153, 'n_estimators': 116}


# Hasil

*   evaluasi : 0.9947275922671354
*   max_feature : 0.25411986663039127
*   min_samples_split : 0.013158723180134153
*   n_estimators : 116



In [8]:
# copy data
data2=data

In [9]:
# import support vector classifier 
# "Support Vector Classifier"
from sklearn.svm import SVC

# inisiasi model
svclassifier = SVC(kernel='linear')

model = svclassifier.fit(X, y) 
print('w = ',svclassifier.coef_)
print('b = ',svclassifier.intercept_)
print('Indices of support vectors = ', svclassifier.support_)
print('Support vectors = ', svclassifier.support_vectors_)
print('Number of support vectors for each class = ', svclassifier.n_support_)
print('Coefficients of the support vector in the decision function = ', np.abs(svclassifier.dual_coef_))

# Evaluasi dengan score
score = model.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {rsc.best_params_}')

w =  [[-1.16914932e+00 -3.98656611e-02  6.09855130e-02  1.17894530e-03
   2.92607137e-01  2.45077126e-01  6.83704900e-01  4.18968471e-01
   4.15846869e-01  1.74636323e-02  4.56971586e-02 -9.85253852e-01
  -4.36534725e-01  8.70137578e-02  5.47753992e-02 -7.61972297e-02
   7.62755322e-02  6.67075458e-02  5.06873585e-02 -1.99392717e-02
   8.77762454e-02  2.32891390e-01  1.01172830e-01  2.25970843e-03
   5.18797742e-01  5.85869547e-01  1.59831544e+00  7.58670399e-01
   1.02773741e+00  6.37112401e-02]]
b =  [-11.25150949]
Indices of support vectors =  [ 90 112 147 157 204 208 225 228 238 290 291 340 347 363 377 410 413 421
 448 455 465 466 476 481 491 500 541 542 560   3   5  13  38  39  40  41
  44  47  73  86  91  99 135 146 194 205 213 215 255 263 277 297 329 385
 479 489 536]
Support vectors =  [[1.462e+01 2.402e+01 9.457e+01 ... 6.946e-02 2.522e-01 7.246e-02]
 [1.426e+01 1.965e+01 9.783e+01 ... 1.505e-01 2.398e-01 1.082e-01]
 [1.495e+01 1.877e+01 9.784e+01 ... 8.405e-02 2.852e-01 9.218

In [10]:
# inisiasi model
svclassifier = SVC(kernel='poly', degree=8)

model = svclassifier.fit(X, y)

# Evaluasi dengan score
score = model.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {rsc.best_params_}')

Hasil evaluasi: 0.9050966608084359
Konfigurasi hyperparameter: {'max_features': 0.25411986663039127, 'min_samples_split': 0.013158723180134153, 'n_estimators': 116}


In [11]:
# inisiasi model
svclassifier = SVC(kernel='rbf')

model = svclassifier.fit(X, y)

# Evaluasi dengan score
score = model.score(X,y)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {rsc.best_params_}')

Hasil evaluasi: 0.9226713532513181
Konfigurasi hyperparameter: {'max_features': 0.25411986663039127, 'min_samples_split': 0.013158723180134153, 'n_estimators': 116}
